abci/example/example_test.go

```diff
@@ -77,51 +76,24 @@ func testStream(ctx context.Context, t *testing.T, logger log.Logger, app types.
 	err = client.Start(ctx)
 	require.NoError(t, err)
 
-	done := make(chan struct{})
-	counter := 0
-	client.SetResponseCallback(func(req *types.Request, res *types.Response) {
-		// Process response
-		switch r := res.Value.(type) {
-		case *types.Response_DeliverTx:
-			counter++
-			if r.DeliverTx.Code != code.CodeTypeOK {
-				t.Error("DeliverTx failed with ret_code", r.DeliverTx.Code)
-			}
-			if counter > numDeliverTxs {
-				t.Fatalf("Too many DeliverTx responses. Got %d, expected %d", counter, numDeliverTxs)
-			}
-			if counter == numDeliverTxs {
-				go func() {
-					time.Sleep(time.Second * 1) // Wait for a bit to allow counter overflow
-					close(done)
-				}()
-				return
-			}
-		case *types.Response_Flush:
-			// ignore
-		default:
-			t.Error("Unexpected response type", reflect.TypeOf(res.Value))
-		}
-	})
-
-	// Write requests
+	// Construct request
+	rfb := types.RequestFinalizeBlock{Txs: make([][]byte, numDeliverTxs)}
 	for counter := 0; counter < numDeliverTxs; counter++ {
-		// Send request
-		_, err = client.DeliverTxAsync(ctx, types.RequestDeliverTx{Tx: []byte("test")})
+		rfb.Txs[counter] = []byte("test")
+	}
+	// Send bulk request (several times)
+	for i := 0; i < 3; i++ {
+		res, err := client.FinalizeBlock(ctx, rfb)
```

Comment:
The original test was testing many asynchronous calls to `DeliverTx`. Of course, asynchronously calling `FinalizeBlock` many times doesn't make sense now. So I decided to turn the test into a bulk test for `FinalizeBlock`.
You got a point in that, on the field, apps won't get `FinalizeBlock` calls in a row (there will at least be calls to `ProcessProposal`, `ExtendVote`, etc). But the same could be sais about the original test case: before calling `DeliverTx` asynchronously many times, you need `BeginBlock`,  and just after you need `EndBlock`.
I followed the same approach with my version of the test with the thought that "it's a UT, so it's _partially_ testing the API"
Now, having said all this, if you strongly feel I need to add some `ProcessProposal` and `ExtendVote` calls to better reflect reality, I can do it

Reply:
Yeah, I definitely see that the original test was not calling all of the ABCI methods so it was a little bit off before. The test probably could do with a little bit of work to make it resemble how we expect it to actually be called, but we don't have to fix every problem we encounter.

I think we can just call `FinalizeBlock` one time since this will test that the application is able to actually update its internal state, which I believe is really what we want to ensure. I'm not certain we need to actually call it multiple times but perhaps there is something I'm missing.

Reply:
Yeah that's a good tradeoff. I'll change that